src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php

```diff
@@ -828,4 +852,23 @@ private function validateInGroup($value, $cacheKey, MetadataInterface $metadata,
             $validator->validate($value, $constraint);
         }
     }
+
+    /**
+     * @param object $object
+     *
+     * @return string
+     */
+    private function generateCacheKey($object)
+    {
+        $this->cachedObjectsRefs[] = $object;
+
+        return spl_object_hash($object);
+    }
+
+    private function cleanUpObjectRefs()
```

Comment:
shouldn't calls to this `cleanUpObjectRefs` happen in a `finally` block instead, to be sure we also record the end of the callstack when some called code triggers an exception ?

Reply:
In fact keeping the object references in the validator instance does not work as expected as new validator instances can be created in one validation run. I decided to move the reference handling to the execution context instead which also saves us from worrying about have to clean up the stored references instead.